### PR TITLE
`ctx` project takeover

### PR DESCRIPTION
**Pull request by ewdurbin**

### CODE REVIEW

**Review by dstufft — APPROVED** [review]

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+Performing this analysis on an ongoing basis and freezing accounts with expired
+or near expiration domains is a potential mitigation that could protect absent
+maintainers in the future, at the cost of increased support burden on the team
+of PyPI moderators and admins.
+
+Further Reccomendations
+=======================
+
+We also advise PyPI users to enable multi factor authentication on their PyPI
+accounts following the references at https://pypi.org/help/#twofa
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:173]:
```suggestion
We also advise all PyPI users, but especially project maintainers, to enable
multi factor authentication on their PyPI accounts following the references at
https://pypi.org/help/#twofa
```

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+The ``ctx`` hosted project on PyPI was taken over via user account compromise
+and replaced with a malicious project which contained runtime code which
+collected the content of ``os.environ.items()``` when instantiating ``Ctx``
+objects. The captured environment variables were sent as a base64 encoded query
+parameter to a heroku application running at ``https://anti-theft-web.herokuapp.com``.
+
+Between 2022-05-14T19:18:36Z and 2022-05-24T10:07:17Z the following release
+below files were hosted by PyPI at various times containing this malicious
+payload
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:16]:
```suggestion
payload.
```

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+If you installed the package between May 14, 2022 and May 24, 2022, and your
+environment variables contain sensitive data like passwords and API keys (like
+AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY), we advise you rotate your
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:20]:
```suggestion
``AWS_ACCESS_KEY_ID`` and ``AWS_SECRET_ACCESS_KEY``), we advise you rotate your
```

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 4]
```diff
@@ -0,0 +1,194 @@
+=========================================================
+Account Takeover and Malicious Replacement of ctx Project
+=========================================================
+
+Summary
+=======
+
+The ``ctx`` hosted project on PyPI was taken over via user account compromise
+and replaced with a malicious project which contained runtime code which
+collected the content of ``os.environ.items()``` when instantiating ``Ctx``
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:10]:
```suggestion
collected the content of ``os.environ.items()`` when instantiating ``Ctx``
```

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+With the malicious nature of all release files confirmed, the PyPI
+administrator used existing tools to:
+
+* Remove the project, all releases, and all release files from the index
+* Simultaneously prohibit the name ``ctx`` from being re-registered without
+  admin intervention
+* Freeze the compromised user account of the owner
+
+The activity log of the user, action log on the project, metadata for all
+historical uploads---including malicious---, archives of the files, and their
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:112]:
```suggestion
historical uploads (including malicious), archives of the files, and their
```

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+complies with that contract will need to be developed.
+
+
+Impact Assessment
+=================
+
+The ``ctx`` project was registered and uploaded to PyPI in 2014. According to
+Libraries.io, the project on PyPI that declares it as a dependency is
+``context-engine``. No known repositories that Libraries.io analyzes declares
+``ctx`` as a dependency.
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:136]: https://deps.dev/pypi/ctx/0.1.2/dependents confirms this as well.

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+Potential Future Mitigation
+===========================
+
+Domain takeovers are a known attack vector for compromising individual user
+accounts on PyPI. PyPI administrators have responded to reports in the past of
+publicly visible email addresses associated with project metadata containing
+expired domains, which happened to match the domains of owner user accounts for
+projects.
+
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:163]:
```suggestion

Currently, PyPI has some protections in place for expired email domains: if
PyPI sends an email to a user's email address, and that email bounces, PyPI
will disable the verified status of that email. As password resets require
verified email addresses, an attacker would be unable to use the expired domain
to gain access to the account. However, this depends on PyPI sending an email
to the expired domain in the time period between expiry and an attacker
attempting a takeover.

```

File: pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+or near expiration domains is a potential mitigation that could protect absent
+maintainers in the future, at the cost of increased support burden on the team
+of PyPI moderators and admins.
+
+Further Reccomendations
+=======================
+
+We also advise PyPI users to enable multi factor authentication on their PyPI
+accounts following the references at https://pypi.org/help/#twofa
+
```
Comment (di) [inline pypi-vuln/index-2022-05-24-ctx-domain-takeover.rst:174]:
```suggestion

Additionally, version-pinning and using `hash checking mode
<https://pip.pypa.io/en/stable/topics/secure-installs/#hash-checking-mode`_
would prevent this attack, which depends on users automatically upgrading to
the latest available version at install-time.

```

**Pull request merged**